features/map.go

```diff
@@ -129,37 +147,61 @@ func haveMapType(mt ebpf.MapType) error {
 		return err
 	}
 
+	err = checkMapCreation(mt, createMapTypeAttr(mt))
+
+	mc.mapTypes[mt] = err
+
+	return err
+}
+
+func haveMapFlags(mt ebpf.MapType, flags uint32) error {
+	cacheEntry := mapFlagCacheEntry{mt, flags}
+
+	mc.Lock()
+	defer mc.Unlock()
+	err, ok := mc.mapFlags[cacheEntry]
+	if ok {
+		return err
+	}
+
+	attr := createMapTypeAttr(mt)
+	attr.MapFlags |= flags
+	err = checkMapCreation(mt, attr)
+
+	mc.mapFlags[cacheEntry] = err
+
+	return err
+}
+
+func checkMapCreation(mt ebpf.MapType, attr *sys.MapCreateAttr) error {
 	fd, err := sys.MapCreate(createMapTypeAttr(mt))
 
 	switch {
 	// For nested and storage map types we accept EBADF as indicator that these maps are supported
 	case errors.Is(err, unix.EBADF):
 		if isMapOfMaps(mt) || isStorageMap(mt) {
-			err = nil
+			return nil
 		}
 
 	// EINVAL occurs when attempting to create a map with an unknown type.
 	// E2BIG occurs when MapCreateAttr contains non-zero bytes past the end
 	// of the struct known by the running kernel, meaning the kernel is too old
 	// to support the given map type.
 	case errors.Is(err, unix.EINVAL), errors.Is(err, unix.E2BIG):
-		err = ebpf.ErrNotSupported
+		return ebpf.ErrNotSupported
 
 	// EPERM is kept as-is and is not converted or wrapped.
 	case errors.Is(err, unix.EPERM):
-		break
+		return nil
 
 	// Wrap unexpected errors.
 	case err != nil:
-		err = fmt.Errorf("unexpected error during feature probe: %w", err)
+		return fmt.Errorf("unexpected error during feature probe: %w", err)
 
-	default:
-		fd.Close()
 	}
 
-	mc.mapTypes[mt] = err
```

Comment:
Removing this makes the caching obsolete and performs the check every time. Is there a particular reason to do so?

Reply:
This is done [here](https://github.com/cilium/ebpf/pull/673/files#diff-df55ec79721c085fabb3b83f2498ad96eff807311889d298b09693c4e1cda8a4R152) now. Because this part is extracted into a separate function